src/WebJobs.Script.WebHost/Diagnostics/MetricsEventManager.cs

```diff
@@ -469,66 +462,86 @@ internal void Flush()
             {
                 double executionTimespan = 0;
-                if (runningFunctionInfo.ExecutionStage == ExecutionStage.Finished)
+                ExecutionStage executionStage;
+                if (!runningFunctionInfo.Completed)
                 {
-                    executionTimespan = (runningFunctionInfo.EndTime - runningFunctionInfo.StartTime).TotalMilliseconds;
+                    executionStage = ExecutionStage.InProgress;
+                    executionTimespan = (currentTime - runningFunctionInfo.Timestamp).TotalMilliseconds;
                 }
                 else
                 {
-                    executionTimespan = (currentTime - runningFunctionInfo.StartTime).TotalMilliseconds;
+                    // regardless of the actual Failed/Succeeded status, we always raise the final event
+                    // with stage Finished
+                    executionStage = ExecutionStage.Finished;
+                    executionTimespan = runningFunctionInfo.Duration.TotalMilliseconds;
                 }
 
                 MetricsEventGenerator.LogFunctionExecutionEvent(
                     _executionId,
                     _appServiceOptions.CurrentValue.AppName,
                     concurrency,
-                    runningFunctionInfo.Name,
+                    runningFunctionInfo.FunctionMetadata.Name,
                     runningFunctionInfo.InvocationId.ToString(),
-                    runningFunctionInfo.ExecutionStage.ToString(),
+                    executionStage.ToString(),
                     (long)executionTimespan,
                     runningFunctionInfo.Success);
 
                 if (_metricsPublisher != null)
                 {
                     _metricsPublisher.AddFunctionExecutionActivity(
-                        runningFunctionInfo.Name,
+                        runningFunctionInfo.FunctionMetadata.Name,
                         runningFunctionInfo.InvocationId.ToString(),
                         concurrency,
-                        runningFunctionInfo.ExecutionStage.ToString(),
+                        executionStage.ToString(),
                         runningFunctionInfo.Success,
                         (long)executionTimespan,
                         _executionId,
                         currentTime,
-                        runningFunctionInfo.StartTime);
+                        runningFunctionInfo.Timestamp);
                 }
 
                 if (_linuxContainerActivityPublisher != null)
                 {
-                    var activity = new ContainerFunctionExecutionActivity(DateTime.UtcNow, runningFunctionInfo.Name,
-                        runningFunctionInfo.ExecutionStage, runningFunctionInfo.TriggerType,
+                    var triggerType = runningFunctionInfo.FunctionMetadata.Trigger?.Type;
+                    var activity = new ContainerFunctionExecutionActivity(DateTime.UtcNow, runningFunctionInfo.FunctionMetadata.Name,
```

Comment:
This is a huge nit (and it was here before) -- but should this use the passed-in `currentTime` instead of `UtcNow`?

Reply:
You may be right, but I didn't want to change this here. In the future, when adding/changing tests we might change this to use the passed in time as we do above.